--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,3 +1,4 @@
+cmake_minimum_required(VERSION 3.10)
 PROJECT (cdrkit C)
 SUBDIRS(include genisoimage wodim libedc libhfs_iso libparanoia icedax libusal librols libunls readom netscsid 3rd-party/dirsplit)
 
--- a/genisoimage/genisoimage.h
+++ b/genisoimage/genisoimage.h
@@ -376,7 +376,7 @@
 extern int	split_SL_component;
 extern int	split_SL_field;
 extern char	*trans_tbl;
-char		*outfile;
+extern char	*outfile;
 
 #define	JMAX		64	/* maximum Joliet file name length (spec) */
 #define	JLONGMAX	103	/* out of spec Joliet file name length */
