--- a/modules.d/45plymouth/module-setup.sh
+++ b/modules.d/45plymouth/module-setup.sh
@@ -19,9 +19,7 @@ read_setting_from_config() {
 
 use_simpledrm() {
     local config use_simpledrm config_dirs=()
-    if [[ $hostonly ]]; then
-        config_dirs+=("${dracutsysrootdir-}/etc/plymouth/plymouthd.conf")
-    fi
+    config_dirs+=("${dracutsysrootdir-}/etc/plymouth/plymouthd.conf")
     for config in "${config_dirs[@]}" "${dracutsysrootdir-}/usr/share/plymouth/plymouthd.defaults"; do
         use_simpledrm=$(read_setting_from_config UseSimpledrm "$config")
         if [[ $use_simpledrm -ge 1 ]]; then
