diff '--color=auto' -apruN a/stdlib/abort.c b/stdlib/abort.c
--- a/stdlib/abort.c	2025-07-28 19:22:55.000000000 +0100
+++ b/stdlib/abort.c	2025-09-05 13:19:23.644737544 +0100
@@ -19,6 +19,7 @@
 #include <internal-signals.h>
 #include <libc-lock.h>
 #include <pthreadP.h>
+#include <string.h>
 #include <unistd.h>
 
 /* Try to get a machine dependent instruction which will make the
@@ -42,7 +43,8 @@ __libc_rwlock_define_initialized (static
 void
 __abort_fork_reset_child (void)
 {
-  __libc_rwlock_init (lock);
+  __libc_rwlock_define_initialized (, reset_lock);
+  memcpy (&lock, &reset_lock, sizeof (lock));
 }
 
 void
