--- a/cube/CMakeLists.txt
+++ b/cube/CMakeLists.txt
@@ -308,13 +308,13 @@ if (ANDROID)
 endif()
 
 if (XCB_LINK_LIBRARIES)
-    target_compile_definitions(vkcube PRIVATE "XCB_LIBRARY=\"${XCB_LINK_LIBRARIES}\"")
+    target_compile_definitions(vkcube PUBLIC "XCB_LIBRARY=\"${pkgcfg_lib_XCB_xcb}\"")
 endif()
 if (X11_LINK_LIBRARIES)
-    target_compile_definitions(vkcube PRIVATE "XLIB_LIBRARY=\"${X11_LINK_LIBRARIES}\"")
+    target_compile_definitions(vkcube PUBLIC "XLIB_LIBRARY=\"${pkgcfg_lib_X11_X11}\"")
 endif()
 if (WAYLAND_CLIENT_LINK_LIBRARIES)
-    target_compile_definitions(vkcube PRIVATE "WAYLAND_LIBRARY=\"${WAYLAND_CLIENT_LINK_LIBRARIES}\"")
+    target_compile_definitions(vkcube PUBLIC "WAYLAND_LIBRARY=\"${pkgcfg_lib_WAYLAND_CLIENT_wayland-client}\"")
 endif()
 
 # ----------------------------------------------------------------------------
@@ -360,14 +360,14 @@ target_include_directories(vkcubepp PRIVATE .)
 target_compile_definitions(vkcubepp PRIVATE ${ENABLED_CUBE_PLATFORMS})
 target_link_libraries(vkcubepp  Vulkan::Headers volk::volk_headers)
 
-if (XCB_LINK_LIBRARIES )
-    target_compile_definitions(vkcubepp PUBLIC "XCB_LIBRARY=\"${XCB_LINK_LIBRARIES}\"")
+if (XCB_LINK_LIBRARIES)
+    target_compile_definitions(vkcubepp PUBLIC "XCB_LIBRARY=\"${pkgcfg_lib_XCB_xcb}\"")
 endif()
 if (X11_LINK_LIBRARIES)
-    target_compile_definitions(vkcubepp PUBLIC "XLIB_LIBRARY=\"${X11_LINK_LIBRARIES}\"")
+    target_compile_definitions(vkcubepp PUBLIC "XLIB_LIBRARY=\"${pkgcfg_lib_X11_X11}\"")
 endif()
 if (WAYLAND_CLIENT_LINK_LIBRARIES)
-    target_compile_definitions(vkcubepp PUBLIC "WAYLAND_LIBRARY=\"${WAYLAND_CLIENT_LINK_LIBRARIES}\"")
+    target_compile_definitions(vkcubepp PUBLIC "WAYLAND_LIBRARY=\"${pkgcfg_lib_WAYLAND_CLIENT_wayland-client}\"")
 endif()
 
 if(APPLE)
